File: php/kaltura-api-notifications-handler-php/lib/Kaltura/Notification/Processor.php
<?php

/**
 * @namespace
 */
namespace Kaltura\Notification;

/**
 * Kaltura Notification Processor class
 *
 * @package Kaltura
 * @subpackage Notification
 */
class Processor
{

	/**
	 * @var int
	 */
	public $handlers = array();
	
	/**
	 * @var \Kaltura\Notification\Client
	 */
	private $notificationClient = null;

	/**
	 * @var \Kaltura\Client\Client
	 */
	private $client = null;

	/**
 	 * Constructor, instanciate a \Kaltura\Notification\Processor object
 	 *
 	 * @param Array, $params: array of parameters. Could be POST, GET or any array from user input
 	 * @param Boolean, $validateClientSignature: whether or not validate the notification signature
 	 * @param String, $adminSecret: Kaltura admin secret (required if validating the notification signature)
	 *
	 * @see \Kaltura\Notification\Client
	 * 
	 * @return @void
 	 */
	public function __construct($params, $validateClientSignature = true, $adminSecret = null) {
		$this->notificationClient = new Client($params, $validateClientSignature, $adminSecret);
	}

	/**
 	 * Add a notification handler to the notification processor
 	 *
 	 * @param \Kaltura\Notification\Handler, $notificationHandler: notification handler to add
 	 * 
 	 * @return \Kaltura\Notification\Handler: handler added
 	 */
	public function addHandler(\Kaltura\Notification\Handler $notificationHandler) {
		$notificationTypes = $notificationHandler->getTypes();

		if ($notificationTypes) {
			foreach ($notificationTypes as $notificationType) {
				$this->handlers[$notificationType][] = $notificationHandler;
			}
		} else {
			$this->handlers[\Kaltura\Notification\Handler::HANDLER_TYPE_ALWAYS][] = $notificationHandler;
		}

		return $notificationHandler;
	}

	/**
 	 * Add a notification handler to the notification processor
 	 *
 	 * @param Array, $notificationHandlers: array of \Kaltura\Notification\Handler objects
 	 * 
 	 * @return \Kaltura\Notification\Handler: handler added
 	 */
	public function addHandlers($notificationHandlers) {
		if (count($notificationHandlers)) {
			foreach ($notificationHandlers as $notificationHandler) {
				$this->addHandler($notificationHandler);
			}
		}

		return $notificationHandlers;
	}

	/**
 	 * Process the notification handlers in orders:
 	 *		- pre-process handlers
 	 *		- handler corresponding to the notification type currently triggered
 	 * 	- handler that should be always executed regardless of the notification type
 	 * 	- post-process handlers
 	 * 
 	 * Note that if there are more than 1 handler per group type, they are executed in the
 	 * order they have been set up
 	 * @see executeHandlersByType
 	 *
 	 * @return Array: result
 	 */
	public function execute() {
		$sequence = array(
			\Kaltura\Notification\Handler::HANDLER_TYPE_PRE_PROCESS,
			$this->notificationClient->type,
			\Kaltura\Notification\Handler::HANDLER_TYPE_ALWAYS_PROCESS,
			\Kaltura\Notification\Handler::HANDLER_TYPE_POST_PROCESS,
		);

		$result = array();

		foreach ($sequence as $handlerType) {
			$result[$handlerType] = $this->executeHandlersByType($handlerType);
		}

		return $result;
	}

	/**
 	 * Process a given notification handler group in the order they have been set up
 	 *
 	 * @return Strin|Array, $notificationType: notification type(s) that you want to process
 	 *
 	 * @return Array: result of handler executions
 	 */
	public function executeHandlersByType($type) {
		$result = array();

		if (isset($this->handlers[$type])) {
			foreach ($this->notificationClient->data as $notificationData) {
				foreach ($this->handlers[$type] as $handler) {
					$result[] = $handler->execute($notificationData);
				}
			}
		}

		return $result;
	}
}

File: java/NotificationHandler/src/lib/Kaltura/Notification/Processor.java
package lib.Kaltura.Notification;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

import lib.Kaltura.Notification.Types.HandleEventType;
import lib.Kaltura.Notification.Types.HandlerProcessType;
import lib.Kaltura.Notification.Types.NotificationType;
import lib.Kaltura.Output.Console;

/**
 * This is the Notification processor.
 * To this class, handlers register to handle notification events. 
 * Once a notification event arrives - This class is responsible to trigger the matching handlers.
 */
public class Processor
{
	/** Mapping between triggering events to the handlers that should be fired.*/
	private Map<HandleEventType, List<NotificationHandlerAbs>> handlers = new HashMap<HandleEventType, List<NotificationHandlerAbs>>();
	/** Console for output */
	private Console console;
	
	public Processor(Console console) {
		this.console = console;
	}
	
	/**
 	 * Add a notification handler to the notification processor
 	 * @param notificationHandler The notification handler to add
 	 */
	public void addHandler(NotificationHandlerAbs notificationHandler) {
		Set<HandleEventType> notificationTypes = notificationHandler.getTypes();

		if (notificationTypes.isEmpty()) {
			addHandlerByType(notificationHandler, HandlerProcessType.ALWAYS_PROCESS);
		} else {
			for (HandleEventType handlerType : notificationTypes) {
				addHandlerByType(notificationHandler, handlerType);
			}
		}
	}
	
	/**
	 * Add a notification handler to a specific notification event
	 * @param notificationHandler The notification handler to add
	 * @param type The event type to register the handler to
	 */
	public void addHandlerByType(NotificationHandlerAbs notificationHandler, HandleEventType type) {
		this.console.write("Processor: Added handler (" + notificationHandler.getClass().getSimpleName() + ") to handle " + type);
		if(!this.handlers.containsKey(type)) 
			this.handlers.put(type, new ArrayList<NotificationHandlerAbs>());
		this.handlers.get(type).add(notificationHandler);
	}

	/**
 	 * Process the notification handlers in orders.
 	 * 
 	 * Note that if there are more than 1 handler per group type, they are executed in the
 	 * order they have been set up
 	 *
 	 * @return Array: result
	 * @throws java.lang.Exception 
 	 */
	public void execute(Notification notification) throws java.lang.Exception {
		
		Map<Integer, Map<String, String>> multiNotification = notification.getData();
		
		executeByType(multiNotification, HandlerProcessType.PRE_PROCESS);
		executeByType(multiNotification, null);
		executeByType(multiNotification, HandlerProcessType.ALWAYS_PROCESS);
		executeByType(multiNotification, HandlerProcessType.POST_PROCESS);
	}

	/**
	 * Handles a specific handlers type. 
	 * If not given, chooses by the request type
	 * @param multiNotification
	 * @param type
	 * @throws Exception
	 */
	protected void executeByType(Map<Integer, Map<String, String>> multiNotification, HandleEventType type) throws Exception {
		for (Map<String, String> request : multiNotification.values()) {
			if(type == null)
				type = NotificationType.getType(request.get("notification_type"));
			if(!handlers.containsKey(type))
				continue;
			
			for (NotificationHandlerAbs handler : handlers.get(type)) {
				handler.execute(request);
			}
		}
	}
}
